portal-impl/src/com/liferay/portal/json/jabsorb/serializer/LiferayJSONSerializer.java

```diff
@@ -105,10 +113,80 @@ else if (_log.isWarnEnabled()) {
 					"Unable to get class " + className, exception);
 			}
 		}
+		else if (object instanceof JSONArray) {
+			JSONArray jsonArray = (JSONArray)object;
+
+			if (jsonArray.length() == 0) {
+				return Object[].class;
+			}
+
+			Class<?> compClazz;
+
+			try {
+				Object jsonArrayFirstItem = jsonArray.get(0);
+
+				compClazz = getClassFromHint(jsonArrayFirstItem);
+
+				if (Objects.equals(compClazz, Integer.class) &&
+					((Integer)jsonArrayFirstItem == 0)) {
+
+					Set<Class<?>> clazzSet = new HashSet<>();
```

Comment:
@slnn The case you need to handle is only `Integer` vs `Long`, right?

You don't need a set, and a `_getWidestType()` method with a dictionary. Just check if every element is `Integer`. If `Long` comes out once, you can break and return the type of long array; otherwise, return the type of int array.

Reply:
Done